esp/services/ws_topology/ws_topologyService.cpp

```diff
@@ -539,64 +540,118 @@ bool CWsTopologyEx::findTimestampAndLT(const char *logname, IFile* rFile, ReadLo
 
     char* pTr = (char*) dataBuffer.str();
     readLogReq.ltBytes = findLineTerminator(pTr, bytesRead);
+    readLogMessageFields(pTr, bytesRead, readLogReq);
 
-    if (rFile->size() < 28) //lineID + timestamp
-        return false;//no timestamp in this log
-    if (!readLogTime(pTr, 9, 19, latestLogTime))
+    if (rFile->size() < (readLogReq.logDatePos + LOG_DATE_TIME_LENGTH))
         return false;//no timestamp in this log
+    if ((readLogReq.logfields == 0) && !readLogTime(pTr, readLogReq.logDatePos, LOG_DATE_TIME_LENGTH, latestLogTime))
+        return false;//Only legacy log file (readLogReq.logfields = 0) may not have timestamp.
 
     if ((readLogReq.filterType != GLOLastNHours) && (readLogReq.filterType != GLOTimeRange))
         return true;//Not interested in the last timestamp this time
 
-    //Search the last chuck to find out the latest timestamp
-    if (readSize < fileSize)
+    return readLastLogTime(logname, rIO, fileSize, readSize, readLogReq, latestLogTime);
+}
+
+void CWsTopologyEx::readLogMessageFields(char* contentStart, size32_t bytesRemaining, ReadLog& readLogReq)
+{
+    //Find out the first log line
+    char* finger = contentStart;
+    while (bytesRemaining > 0)
     {
-        bytesRead = rIO->read(fileSize - readSize, readSize, dataBuffer.clear().reserve(readSize));
-        if (bytesRead != readSize)
-            throw MakeStringException(ECLWATCH_CANNOT_READ_FILE, "Failed to read file %s.", logname);
-        pTr = (char*) dataBuffer.str();
+        if (isLineTerminator(finger, bytesRemaining, readLogReq.ltBytes))
+            break;
+
+        finger++;
+        bytesRemaining--;
     }
 
-    //skip the first line which may not contain the entire log line
-    StringBuffer logLine;
-    size32_t bytesRemaining = bytesRead;
-    bool hasAnotherLine = false;
-    pTr = readALogLine(pTr, bytesRemaining, readLogReq.ltBytes, logLine, hasAnotherLine);
-    if (!hasAnotherLine || (bytesRemaining < 28))
-        return true;
+    if (finger > contentStart)
+    {//Try to read LogMessageFields from the first log line
+        StringBuffer logLine;
+        logLine.append(finger - contentStart, contentStart);
+        readLogReq.logfields = getMessageFieldsFromHeader(logLine);
+        if (readLogReq.logfields > 0)
```

Comment:
If readLogReq.logfields==0, then readLogReq.logfields should be set to MSGFIELD_LEGACY to use the legacy format